File: Day 11/Valid Sudoku (Set  validation).py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [[] for _ in range(9)]
        cols = [[] for _ in range(9)]
        boxs = [[] for _ in range(9)]

        for r in range(9):
            for c in range(9):
                val = board[r][c]
                if val == ".":
                    continue

                if val in rows[r] or val in cols[c] or val in boxs[(r//3)*3 + (c//3)]:
                    return False
                rows[r].append(val)
                cols[c].append(val)
                boxs[(r//3)*3 + (c//3)].append(val)
        return True

File: Day 11/Valid Sudoku (Set  validation).cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<vector<char>> rows(9), cols(9), boxes(9);

        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                char ch = board[r][c];
                if (ch == '.') continue;

                int b = (r/3)*3 + (c/3); // 0..8

                // duplicate check via linear search (size ≤ 9, so fine)
                if (find(rows[r].begin(), rows[r].end(), ch) != rows[r].end()) return false;
                if (find(cols[c].begin(), cols[c].end(), ch) != cols[c].end()) return false;
                if (find(boxes[b].begin(), boxes[b].end(), ch) != boxes[b].end()) return false;

                rows[r].push_back(ch);
                cols[c].push_back(ch);
                boxes[b].push_back(ch);
            }
        }
        return true;
    }
};
